### ISSUE TITLE
test: run official bash test suite against brush

### ISSUE THREAD

**Issue by 39555** [reubeno/brush#264]
I read about [oils](https://github.com/oils-for-unix/oils) and its extensive https://github.com/oils-for-unix/oils/wiki/Spec-Tests. It is similar to brush yaml files and contains a lot of scripts from real-world codebases. Maybe we could use it directly or auto-format it into YAML files.

The stats for different shells can be found here https://www.oilshell.org/release/0.7.pre11/test/spec.wwz/for-expr.html

**Comment by joshtriplett** [comment]
You may also want to consider running the [testsuite for bash itself](https://git.savannah.gnu.org/cgit/bash.git/tree/tests). (You wouldn't want to integrate that directly in this repository, since it's part of bash, and uses the same license as bash.)

**Comment by reubeno** [comment]
> You may also want to consider running the [testsuite for bash itself](https://git.savannah.gnu.org/cgit/bash.git/tree/tests). (You wouldn't want to integrate that directly in this repository, since it's part of bash, and uses the same license as bash.)

That's a great point, @joshtriplett; thanks for the suggestion.

As you hint, I've been intentional in avoiding the `bash` sources + repo for licensing reasons, but you're right that we could look into how to separately find a way to run its tests. (We've done something similar in running `bash-completion` integration tests using `brush`; they've been surprisingly insightful :smile:.) I'd also like to separately explore getting a temporary license to run official POSIX conformance tests.

**Comment by RGBCube** [comment]
I do not think you need your tool to be GPL licensed to run a GPL test suite. Just don't vendor or redistribute the code as MIT (or whatever).

If that was not possible, you wouldn't be able to run non-GPL programs on Linux, which is a GPL kernel. Treating it as a platform is okay.

**Comment by oech3** [comment]
uutils/coreutils customizes GNU test suite at its repo and runs it at least for.

But if you don't want to add `bash` test to this repo, different people could have a repo fetching `brush` binary and run against bash, and open bug reports from there.

**Comment by reubeno** [comment]
> uutils/coreutils customizes GNU test suite at its repo and runs it at least for.
> 
> But if you don't want to add `bash` test to this repo, different people could have a repo fetching `brush` binary and run against bash, and open bug reports from there.

Agreed. I’m even okay running it as part of PR checks / CI but don’t want to check it in here. We’ve done this with the `bash-completion` test suite, which runs on all PRs (cloned directly from source); that actually caught a lot!

As for the bash tests, there’s actually a primitive script checked into the `scripts` dir of this repo, called something like `run-bash-tests.py` that I’ve used manually to run `bash`’s test suite against bash and brush. One challenge is that the bash test suite is a bit harder to extract automated result metrics from. It’s got the kind of tests that emit nothing when they succeed, warnings when the result requires human review, and textual errors when they fail.

I’d definitely be open to someone extending this py script into a runner that tries to generate a results report that we could generate during PR checks and use to track our progress over time like uutils does against coreutils tests.

There’s also good opportunity to leverage the oils test suites and somehow integrate their execution.
